src/TorchSharp/DisposeScopeManager.cs

```diff
@@ -0,0 +1,306 @@
+            {
+                Exclude(new IDisposable[] { first, second, third }, false);
+                return (first, second, third);
+            }
+
+            public IDisposable[] Exclude(IDisposable[] exclude, bool excludeGlobally)
+            {
+                lock (this) {
+                    foreach (var disposable in exclude) {
+                        if (!Disposables.Contains(disposable)) {
+                            throw new InvalidOperationException("The disposable does not belong to this scope!");
+                        }
+
+                        Disposables.Remove(disposable);
+                        if (!excludeGlobally) {
+                            Singleton.MoveToOuterScope(this, disposable);
+                        }
+                    }
+                }
+
+                return exclude;
+            }
+
+            /// <summary>
+            /// Excludes a set of tensors/disposables from the all dispose scopes, see overloaded methods. See Exclude
+            /// if you wish to move it to the outer dispose scope.
+            /// </summary>
+            public T ExcludeGlobally<T>(T exclude) where T : IDisposable
+            {
+                Exclude(new IDisposable[] { exclude }, true);
+                return exclude;
+            }
+
+            public (T1 first, T2 second) ExcludeGlobally<T1, T2>(T1 first, T2 second)
+                where T1 : IDisposable where T2 : IDisposable
+            {
+                Exclude(new IDisposable[] { first, second }, true);
+                return (first, second);
+            }
+
+            public (T1 first, T2 second, T3 third) ExcludeGlobally<T1, T2, T3>(T1 first, T2 second, T3 third)
+                where T1 : IDisposable where T2 : IDisposable where T3 : IDisposable
+            {
+                Exclude(new IDisposable[] { first, second, third }, true);
+                return (first, second, third);
+            }
+
+            /// <summary>
+            /// Disposes everything currenly in the dispose scope.
+            /// </summary>
+            public void DisposeEverything() => DisposeEverythingBut();
+
+            /// <summary>
+            /// As an intermediate step, you can dispose all the tensors/disposables currently scheduled for dispose, to
+            /// clear up some memory without creating a new scope. Note that this doesn't permanently exclude the
+            /// tensors from disposing, use Exclude for that. Also, excluded tensors don't need to be included
+            /// here.
+            /// </summary>
+            public IDisposable[] DisposeEverythingBut(params IDisposable[] keep)
+            {
+                lock (this) {
+                    foreach (var disposable in keep) {
+                        if (!Disposables.Contains(disposable)) {
+                            throw new InvalidOperationException("The disposable does not belong to this scope!");
```

Comment:
It's potentially quite a benign error to exclude something that isn't actually in the scope. It feels good to check for this, but I don't think it's disastrous to ignore it.

Reply:
Agreed, sometimes I add that kind of tests  to make sure stuff's actually working the way I intended, if I just ignored that kind of "error" when coding could be hiding a bigger problem. But once it works as intended, if someone wanted to Exclude something that wasn't actually ever Included, then that should probably just be ignored.

Reply:
Oh, also, should it still be moved to the outer scope? What if it never belonged to the outer scope, and I ask for it to be ignored, where does it go? I'm thinking we just ignore it and move on, without moving it up?

Reply:
I don't think something that was never in the inner scope should automatically be moved to the outer scope. If, for example, it was excluded globally, it could create some issues to do so. For example, I imagine that the tensors created in the input data preprocessing pipeline will be excluded globally, and you really do not want those accidentally transferred to a parent scope.

Reply:
Agreed